### ISSUE TITLE
Unexpected outcome when using a robust command to specify a filename for `\input`

### ISSUE THREAD

**Issue by jfbu** [latex3/latex2e#528]
## Brief outline of the bug

I don't know is this expected:

```
(./testlatexinput2.tex
LaTeX2e <2020-10-01> patch level 4
L3 programming layer <2021-02-18>
(/usr/local/texlive/2021/texmf-dist/tex/latex/latexbug/latexbug.sty)
(/usr/local/texlive/2021/texmf-dist/tex/latex/base/article.cls
Document Class: article 2020/04/10 v1.4m Standard LaTeX document class
(/usr/local/texlive/2021/texmf-dist/tex/latex/base/size10.clo))
(/usr/local/texlive/2021/texmf-dist/tex/latex/l3backend/l3backend-pdftex.def)
(./testlatexinput2.aux)

! LaTeX Error: File `protectfilename.tex' not found.

Type X to quit or <RETURN> to proceed,
or enter new name. (Default extension: tex)
```

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}

\begin{document}

\DeclareRobustCommand\foo{file}

\input{\foo name}
\end{document}

```

and the directory does contain a file with name `filename`.

## Log file (required) and possibly PDF file  

[testlatexinput2.log](https://github.com/latex3/latex2e/files/6100462/testlatexinput2.log)

**Comment by FrankMittelbach** [comment]
What would you expect to happen?

Basically robust commands do not expand unless you are typesetting, so the fact that there is no need to "protect" the word "file" it is not really possible for LaTeX to look inside the definition and determine what to make of it. So it is rendered harmless somehow.

The actual result is of course rather useless and I agree curious, but not much can be done about that imo. If you use \raggedleft instead of \foo you get
```
! LaTeX Error: File `protectlet\@centercrrightskipz@skipleftskip@flushgluefinal
hyphendemerits=z@parindentz@parfillskipz@skipname.tex' not found.
```
which is even worse. Basically everything remotely looking like a character became one. So getting explicitly rid of the "protect" in the beginning wouldn't help.

Note that with the current -dev release it changes so it now tries to find
```
! LaTeX Error: File `foo name.tex' not found.
! LaTeX Error: File `raggedleft name.tex' not found.
```
which still doesn't make robust commands usable but seems a somewhat more consistent handling to me.

**Comment by jfbu** [comment]
I am asking this in the context of #523. I am trying to understand the rationale for the character per character parsing done, and thought, perhaps they implemented stuff not yet in ltnews. Because in ltnews all I could find is mention of spaces and support of Unicode characters (and perhaps at other locations such as graphics stuff with multiple dots in filename): and for this I don't see why one should go into token per token parsing. So I thought, perhaps something else happened, for example related to robust macros.

But I should have read more carefully https://github.com/latex3/latex2e/issues/523#issuecomment-791917844 where @PhelypeOleinik  explained clearly

> it also does selective macro expansion (protected or robust macros are left alone, whereas everything else is expanded),

I don't know in what way robust macros are presented to LaTeX users and whether for a LaTeX user my mwe would have been expected to work or not. If the idea of a robust macro for the LaTeX user is something which moves along safely in so-called (extremely confusing Lamport terminology) moving arguments and expand only at the right time, then expansion in `\input` feels to me as being "at the right time".

If on the other hand the average LaTeX user is told robust macros expand only in typesetting context, then the fact that they are not allowed in `\input` is expected by average LaTeX user. But in that case, why should parsing of filename not raise an error when encountering such a macro: "a robust macro \foo is not allowed here".

Basically, apart from scanning for such erroneous input and thus intervene before something goes wrong I do not understand why there should be a token per token examination of filenames and not some approach via `\edef` and a search for dots via delimited macros. As a much more precise parsing of filename is done, I thought maybe it has to do with such things as robust macros.

**Comment by PhelypeOleinik** [comment]
As Frank already mentioned, with the upcoming release a LaTeX-robust command like in your example will raise the error:
```none
! LaTeX Error: File `foo name.tex' not found.
```
because in the filename parsing code, `\protect` is set equal to `\string`.

Additionally, in the `expl3` filename parser (used by LaTeX2e since the 2020-10-01 release), if an invalid token is found in the file name, for example `\relax` (more specifically any token not allowed in a `\csname..\endcsname`), as in:
```latex
\input{file\relax name.tex}
```
an error message like this will be raised:
```none
! Undefined control sequence.
<argument> \LaTeX3 error: 
                           \relax invalid in file name. Lost: name.tex.
l. ...  }
```
indicating the offending token, and the tokens lost due to the error.  In that case, the file name is taken as `file` (possibly with an added `.tex` extension).

Regarding the error raised in another issue, this uses a simple `\csname`-based expansion, and a cheap process to detect offending tokens, so it should be as fast and robust as it can be (or as much as I could make it :)

**Comment by FrankMittelbach** [comment]
resolved I would say. It is now as we expect it to be in the -dev release.
